File: problems/C++/05. Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.empty()) {
            return "";
        }
        int st = 0, en = 0;

        for (int i = 0; i < s.length(); ++i) {
            int len1 = expd(s, i, i);
            int len2 = expd(s, i, i + 1);
            int mlen = max(len1, len2);
            if (mlen > en - st) {
                st = i - (mlen - 1) / 2;
                en = i + mlen / 2;
            }
        }
        return s.substr(st, en - st + 1);
    }

private:
    int expd(const string& s, int l, int r) {
        while (l >= 0 && r < s.length() && s[l] == s[r]) {
            l--;
            r++;
        }
        return r - l - 1;
    }
};

File: problems/Rust/05. Longest Palindromic Substring.rs
use std::cmp;

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.is_empty() {
            return String::new();
        }

        let b = s.as_bytes();
        let mut st = 0;
        let mut en = 0;

        for i in 0..b.len() {
            let len1 = Self::expd(b, i as isize, i as isize);
            let len2 = Self::expd(b, i as isize, i as isize + 1);
            let mlen = cmp::max(len1, len2);

            if mlen > en - st {
                st = i - (mlen - 1) / 2;
                en = i + mlen / 2;
            }
        }
        s[st..=en].to_string()
    }

    fn expd(b: &[u8], mut l: isize, mut r: isize) -> usize {
        let n = b.len() as isize;
        if l < 0 || r >= n {
            return 0;
        }
        while l >= 0 && r < n && b[l as usize] == b[r as usize] {
            l -= 1;
            r += 1;
        }
        (r - l - 1) as usize
    }
}